src/server.c

```diff
@@ -5603,10 +5703,18 @@ int prepareForShutdown(int flags) {
             serverLog(LL_WARNING,"Error trying to save the DB, can't exit.");
             if (server.supervised_mode == SUPERVISED_SYSTEMD)
                 redisCommunicateSystemd("STATUS=Error trying to save the DB, can't exit.\n");
-            return C_ERR;
+            goto error;
         }
     }
 
+    /* Simulated failure for debug and test. */
+    if (server.debug_flags & DEBUG_PREVENT_SHUTDOWN) {
+        serverLog(LL_WARNING, "Stopped by DEBUG PREVENT-SHUTDOWN, can't exit.");
+        if (server.supervised_mode == SUPERVISED_SYSTEMD)
+            redisCommunicateSystemd("STATUS=Debug prevents shutdown, can't exit.\n");
+        goto error;
+    }
```

Comment:
I don't know how to do that. Is it reliable?

Another way would be to add an undocumented SHUTDOWN flag (SHUTDOWN DEBUG-FAIL).

Reply:
i don't want to add debug flags to other commands, i rather use the DEBUG command over that.
but i do think that we can use the initial AOFRW to reliably trigger that.
it'll add some 10 lines of test code, but maybe it's better than adding 5 lines of redis code.

Reply:
TBH I don't know what an initial AOFRW is. I don't know much about AOF.

Reply:
I'll explain then (for your benefit).
normally if redis starts configured to persist to AOF, it either loads one, or creates one when redis is empty.
in these cases, every write that is executed is appended to the file, and the user can shutdown redis whenever it likes.

I assume you know what an AOFRW means, but i'll just mention that if the files grows by more and more appends, eventually we start a fork to generate a new one based on the data we have in ram (some form of compaction)

there is an odd situation which happens when redis starts without AOF enabled, and then AOF is enabled at runtime (after redis already has data).
In that case, redis starts an AOFRW process, and responds to the CONFIG SET command with OK (agreeing to a new persistence contract), but the data is not yet "safe", it'll only be safe when the rewrite is completed, and this is why it refuses to terminate in that scenario.

Reply:
Thanks for explaining! So it's basically just to set a config value to yes? But I don't see how we can reliably do this just when the replica catches with the master (just before we send SIGCONT to the replica?) and how we can rely on that this AOFRW takes enough time so that finishShutdown is called exactly during this time. Replicas report their offset only once per second.

Reply:
can't `rdb-key-save-delay` do that? i.e. we'll make the master do an AOFRW that takes forever, then unfreeze the replica and issue a SHUTDOWN, the replica will be done before the AOFRW.
maybe i forgot (or didn't understand) what we wanted to test though.

Reply:
A hidden config trick! Makes sense; I'll try it.